src/atdgen_codec_runtime.ml

```diff
@@ -173,7 +173,36 @@ struct
     else
       Some (decode json)
 
-  let fieldOptional s f = optional (field s f)
+  exception MissingFieldError of string
+
+  (* Define our own version of field which can raise MissingFieldError *)
+  let field key decode json =
+    if
+      Js.typeof json = "object" &&
+      not (Js.Array.isArray json) &&
+      not ((Obj.magic json : 'a Js.null) == Js.null)
+    then begin
+      let dict =
+        (Obj.magic (json : Js.Json.t) : Js.Json.t Js.Dict.t) in
+      match Js.Dict.get dict key with
+      | Some value -> begin
+        try
+          decode value
+        with
+          DecodeError msg -> raise @@ DecodeError (msg ^ "\n\tat field '" ^ key ^ "'")
+        end
+      | None ->
+        raise @@ MissingFieldError ({j|Expected field '$(key)'|j})
+    end
+    else
+      raise @@ DecodeError ("Expected object, got " ^ Js.Json.stringify json)
+
+  (* We don't use Json_decode.optional because it swallows exceptions *)
+  let sane_optional decode json = Some (decode json)
```

Comment:
`try .. catch -> None` should be in sane_optional (also maybe better name? and link to bs-json issue)

Reply:
Renamed to custom_optional and added link to bs-json issue: https://github.com/ahrefs/bs-atdgen-codec-runtime/pull/11/commits/1fb4750c124f7e459776bc6214de2bf952ecf55d#diff-af319977d60814aacb356044a4666058R199